### PR TITLE
Add nonces to Store API endpoints

### LINKED ISSUE #1955: Implement nonces for REST API updates in Cart/Checkout
Even though the write endpoints we are introducing for Cart and Checkout are non-authed, we still need to protect against CSRF attacks using nonces.  This issue is to ensure we implement that before launch.

Note: Woo Core already has [handling for logged out nonce creation](https://github.com/woocommerce/woocommerce/blob/0a351fb2aa340a7380055979d91106fecac87f21/includes/class-wc-session-handler.php#L77-L80).

### PR DESCRIPTION
**Pull request by mikejolley**
I think this is ready for some feedback :) Disables cookie auth for the public Store API and introduces custom nonce headers for PUT POST DELETE endpoints.

The initial nonce is hydrated from settings. The nonce value is `wc_store_api`. This nonce is then stored to a shared store `wc/store/nonce`.

Requests that require nonces use the shared control `apiFetchWithNonce` which does current nonce lookup, and adds the header.

`apiFetchWithNonce` will update the nonce in the store when a response is returned by dispatching the `receiveNonceFromHeaders` action.

Closes #1955 

### How to test the changes in this Pull Request:

1. Test cart/all products block flows.
2. Test POST PUT and DELETE endpoints using a client you'll get 403 errors.
3. Difficult to test via client as you need to set the `X-WC-Store-API-Nonce` header (you can get from page source of cart), basic auth (login as same user as you're browsing cart with), and you need to copy the wordpress-loggedin cookie to your client so that the nonce validation works. Possible but difficult. Thats my only issue with nonces - it makes using a client very difficult but I'm not sure if you can work around that without breaking CSRF protection.

### CODE REVIEW

File: assets/js/blocks/cart-checkout/cart/frontend.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ import renderFrontend from '../../../utils/render-frontend.js';
  * Renders the frontend block within the cart provider.
  */
 const Block = ( { emptyCart, attributes } ) => {
-	const { cartItems, cartIsLoading } = useStoreCart();
+	const { cartItems = [], cartIsLoading = true } = useStoreCart();
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/cart/frontend.js]: For some reason when using apiFetchWithHeaders this code failed and gave a length error instead.

[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/blocks/cart-checkout/cart/frontend.js]: Odd 🤔 , so I tried removing these with the changes I suggested above and I didn't get any console errors. Do you remember if the cart was empty when you were getting the error?

[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/cart/frontend.js]: Maybe it was just an error I introduced. ¯\_(ツ)_/¯

**Comment by mikejolley** [comment]
👍 ok @nerrad. Guess I just enjoy wp.data now :p I wanted something global, I wasn't aware it was that simple to use a middleware.

**Comment by nerrad** [comment]
> Guess I just enjoy wp.data now

Welcome to the club!

>  I wasn't aware it was that simple to use a middleware.

In this case we're able to use it because we're using the `wp.apiFetch` external (which means we can add a property to `apiFetch` which allows us to set the nonce in our controls). If we weren't using the `apiFetch` external, we'd have to wrap it in our own object so we could enhance with the nonce setter.

**Comment by mikejolley** [comment]
Ok thats working. I had to modify `apiFetchWithHeaders` so we can pass more options than just `path` in case that wasn't clear.

File: assets/js/settings/blocks/store-api-nonce.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ */
+const isStoreApiGetRequest = ( options ) => {
+	const url = options.url || options.path;
+	if ( ! url || ! options.method || options.method === 'GET' ) {
+		return false;
+	}
+	return /wc\/store\//.exec( url ) !== null;
+};
+
+/** cache for nonce initialized from hydration  */
```
Comment (nerrad) [inline assets/js/settings/blocks/store-api-nonce.js]: I know this is my example, but we should probably just use `//` for the comment block (and add a full stop at the end of the comment sentence) here.

File: assets/js/settings/blocks/store-api-nonce.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+let nonce = getSetting( 'storeApiNonce' );
+
+const setNonce = ( headers ) => {
+	const newNonce = headers?.get( 'X-WC-Store-API-Nonce' );
+	if ( newNonce ) {
+		nonce = newNonce;
+	}
+};
```
Comment (nerrad) [inline assets/js/settings/blocks/store-api-nonce.js]: This is a public api on `apiFetch.setNonce` so we should probably doc block this (Note: I realize there's lots of code I've merged into master that's a public api without doc blocks during this cycle that still needs it, so if you want to do as a follow-up I understand and support).

File: src/RestApi.php [hunk 4]
```diff
@@ -55,6 +56,37 @@ public static function get_routes_from_namespace( $namespace ) {
...
+		$request_uri = esc_url_raw( wp_unslash( $_SERVER['REQUEST_URI'] ) );
+
+		return false !== strpos( $request_uri, $rest_prefix . 'wc/store' );
+	}
+
+	/**
+	 * The Store API does not require authentication.
+	 *
+	 * @param \WP_Error|mixed $result Error from another authentication handler, null if we should handle it, or another value if not.
+	 * @return WP_Error|mixed|bool
```
Comment (nerrad) [inline src/RestApi.php]: I think this is supposed to be `WP_Error|null|bool` (from the filter definition in WP core).

**Pull request merged**